File: src/patterns/structural/adapter/governmentadapterimpl.nim
# std
import
    std/[
        json
    ]

# internal
import
    patterns/structural/adapter/[
        government
    ]


type
    GovernmentAdapterImpl* = ref object
        government: Government


proc newGovernmentAdapterImpl*(government: Government): GovernmentAdapterImpl =
    return GovernmentAdapterImpl(
        government: government
    )

proc sendGainsInformation*(self: GovernmentAdapterImpl, data: JsonNode): bool =
    return self.government.sendGainsInformation((data["taxPayerId"].getInt().uint(), data["taxPayerGains"].getFloat()))

File: src/patterns/structural/adapter/governmentadapterimpl.go
package adapter

import (
	"encoding/json"
)


type GovernmentAdapterImpl struct {
	government *Government
}


func NewGovernmentAdapterImpl(government *Government) *GovernmentAdapterImpl {
	return &GovernmentAdapterImpl{
		government: government,
	}
}

func (self *GovernmentAdapterImpl) SendGainsInfo(data []byte) bool {
	var jsonValues map[string]interface{}
	json.Unmarshal(data, &jsonValues)
	return self.government.SendGainsInformation(Tuple{TaxPayerId: jsonValues["tax_payer_id"], TaxPayerGains: jsonValues["tax_payer_gains"]})
}